frontend/ui/uimanager.lua

```diff
@@ -628,32 +628,69 @@ end
 --]]
 
 --[[--
-Unschedules a previously scheduled task.
+Unschedules previously scheduled tasks.
 
 In order to unschedule anonymous functions, store a reference.
 
-@func action
-@see scheduleIn
+@func action task to be removed
+@number N The maximum number of tasks to be unscheduled. For a well known number, using N improves speed. If N is omitted all action tasks will be removed.
+@return at least on task removed, number of tasks removed.
 
+@see scheduleIn, unscheduleSoonest
 @usage
 
 self.anonymousFunction = function() self:regularFunction() end
 UIManager:scheduleIn(10.5, self.anonymousFunction)
 UIManager:unschedule(self.anonymousFunction)
 ]]
-function UIManager:unschedule(action)
-    local removed = false
+function UIManager:unschedule(action, N)
+    local removed = 0
+    N = N or math.huge
     for i = #self._task_queue, 1, -1 do
         if self._task_queue[i].action == action then
             table.remove(self._task_queue, i)
-            removed = true
+            removed = removed + 1
+            if removed >= N then
+                return true, removed
+            end
         end
     end
-    return removed
+    return false, removed
 end
 dbg:guard(UIManager, 'unschedule',
     function(self, action) assert(action ~= nil) end)
 
+--[[--
+Unschedules the soonest upcoming N previously scheduled tasks.
```

Comment:
Seeing the presence of this function, I was expecting you were adding at once a bunch of scheduled tasks to do the ramp - so you have a way with this function to cancel some/all of them. Which I would say is not at all the way we do stuff (rather, tasks reschedule themselves, so there's never more than one particular task scheduled).
But it seems you do not do that, and you always call this function with N=1. Right ?
So, why keep this function, which invites to do something not really orthodox ? :)

Reply:
> So, why keep this function, which invites to do something not really orthodox ? :)

No, the `unscheduleSoonest` method is just an optimized `unschedule`. When the number of scheduled task is exactly known, there is no need to go through the whole `UIManager:_task_queue`, you can break after the given number of deletes. 

I have also drilled up the `unschedule` method with the same idea (but with a default of `math.huge` entries to delete, so nothing existing will get broken).

The only difference between `unscheduleSoonest(xxx, 1)` and `unschedule(xxx, 1)` would be, from where the task to delete is searched from. If you know it will come _early_ then use `unscheduleSoonest` else `unschedule`.

In short: 
*) If you don't know the number of tasks, nothing has changed, you use `unschedule(xxx)`.
*) If you know the number of scheduled tasks you can use `unscheduleSoonest(xxx, 1)` and `unschedule(xxx, 1)`, they are both _at least_ as fast as the original `unschedule(xxx)`. If you now if the task comes early use `unscheduleSoonest`.


> I was expecting you were adding at once a bunch of scheduled tasks

No, I am doing that by scheduling exactly one function (`dim_task`, which checks if the deadline is blown (standby, suspend ...) or reached):
- If not, `dim_task` schedules itself again (with the correct predicted time).
- If so, the `ramp_task` is executed and scheduling itself, `dim_task` has run out. (Further taps are discarded.)

If the ramp is over, it runs out too, and we are waiting for an input event. If that comes `dim_task` gets started and taps will be accepted (only the wakeup task gets discarded).

Reply:
> If you know the number of scheduled tasks 

The total nb of scheduled tasks ? (which, well, I can't and shouldn't have to know)
Or the nb of instances of my specific "action" task ? (which should always be no more than one).

Anyway, feels a bit like over engineering :/ and requiring devs to ask themselves some uneeded questions (as we at one point don't have more than a few tasks queued).

Reply:
Agreed, this sounds more like something (possibly) for internal use?

Reply:
Anyway, I can drop `unscheduledSoonest`, but I would really like to keep the optional parameter N in `schedule`.

Reply:
@poire-z 
>  Or the nb of instances of my specific "action" task ? (which should always be no more than one).

Of course the number of the specific "action" task.

Reply:
The only parameter name that would not feel odd to me would be something like "first=true" or "only_one=true", meaning you know there is only one instance of that specific action, and you know the removal doesn't need to loop thru all the queue items.
But then, I think it's always the case :) so this might be the default action (but ok, better safe than sorry).
Is there a place in the code you can think of that could get the same action scheduled multiple times ?
Also, how many actions could you get in the queue during your coding and experimentations, that you felt you needed this function/parameter to _not_ walk the full queue?

Reply:
>  But then, I think it's always the case :) so this might be the default action (but ok, better safe than sorry).

For testing I have added a test to count the unscheduled actions during one `unschedule` and sometimes there have been a count of 2. Did not investigate deeper (as I first wanted to get the `scheduled` modification through).

>The only parameter name that would not feel odd to me would be something like "first=true" or "only_one=true", meaning you know there is only one instance of that specific action, and you know the removal doesn't need to loop thru all the queue items.

"first" seems a bad name, as what is first? The first added, the first in the queue (soonest) the first in the algoritm (=last).

What about two functions? `unschedule(action)` (the current function) and `unscheduleSingle(action)`. Then we can migrate one after the other `unschedule` to `unscheduleSingle`.

> Is there a place in the code you can think of that could get the same action scheduled multiple times?

Yes, autowarmth :) But I will change that (sooner or later, as this is not straight forward).

And to be sure I would go through the whole queue in `onWidgetClose`.

Reply:
There are very few things that might schedule twice, yes (it's either a "bug" or a quirk of event ordering, though).

This feels highly over-engineered, because the task queue should always be *very* small, so breaking early is unlikely to serve a real purpose in practice 99% of the time.

Reply:
Ok then I drop it. (But keep this in an edge in my head for future ;) )

Reply:
(I edited my answer a couple of times, in case people only read the first couple of drafts ;p).